src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -184,10 +130,549 @@ def _validate_certificate_with_key(certificate, private_key, schema_name, verror
             'Please provide a valid csr_id which has a valid CSR filed'
         )
 
-    await middleware.run_in_thread(
-        _validate_certificate_with_key, certificate, private_key, schema_name, verrors
+    await middleware.call(
+        'cryptokey.validate_certificate_with_key', certificate, private_key, schema_name, verrors, passphrase
     )
 
+    key_type = data.get('key_type')
+    if key_type:
+        if key_type != 'EC' and not data.get('key_length'):
+            verrors.add(
+                f'{schema_name}.key_length',
+                'This field is required with RSA based keys'
+            )
+
+
+class CryptoKeyService(Service):
+
+    ec_curve_default = 'BrainpoolP384R1'
+
+    ec_curves = [
+        'BrainpoolP512R1',
+        'BrainpoolP384R1',
+        'BrainpoolP256R1',
+        'SECP256K1'
+    ]
+
+    backend_mappings = {
+        'common_name': 'common',
+        'country_name': 'country',
+        'state_or_province_name': 'state',
+        'locality_name': 'city',
+        'organization_name': 'organization',
+        'organizational_unit_name': 'organizational_unit',
+        'email_address': 'email'
+    }
+
+    class Config:
+        private = True
+
+    async def ec_curves_options(self):
```

Comment:
This is used in the UI, this needs to be publicly exposed.

Also we have some sort of policy to call these *_choices and return a dict (key, value).

Reply:
Perhaps in the crypto service instead.